### PR TITLE
[NEW] VoIP Support for Omnichannel

### PR DESCRIPTION
**Pull request by KevLehman**
## Proposed changes (including videos or screenshots)

- Created VoipService to manage VoIP connections and PBX connection
- Created LivechatVoipService that will handle custom cases for livechat (creating rooms, assigning chats to queue, actions when call is finished, etc)
- Created Basic interfaces to support new services and new model
- Created Endpoints for management interfaces
- Implemented asterisk connector on VoIP service
- Created UI components to show calls incoming and to allow answering/rejecting calls
- Added new settings to control call server/management server connection values
- Added endpoints to associate Omnichannel Agents with PBX Extensions
- Added support for event listening on server side, to get metadata about calls being received/ongoing
- Created new pages to update settings & to see user-extension association
- Created new page to see ongoing calls (and past calls)
- Added support for remote hangup/hold on calls
- Implemented call metrics calculation (hold time, waiting time, talk time)
- Show a notificaiton when call is received

## Further comments

This PR (and a new Fuselage release) are needed to avoid having to pass a random className to TopBar.ToolBox https://github.com/RocketChat/Rocket.Chat.Fuselage/pull/603

How it looks:
- Omnichannel Header component
image
- Call received component
image
- Call room when call's ongoing
image
- Call information panel (from call room)
image
- Call events attached to room
image
- Wrapup when call is finished
image
- Omnichannel contact center (with calls support)
image
- Call information panel (from contact center)
image

### CODE REVIEW

**Comment by KevLehman** [comment]
Updated base branch with latest of develop, again

File: app/livechat/lib/VoipRoomType.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import { IRoom } from '../../../definition/IRoom';
+import { ISubscription } from '../../../definition/ISubscription';
+// @ts-ignore has no exported member
```
Comment (KevLehman) [inline app/livechat/lib/VoipRoomType.ts:3]: Why do we need these ignores?

File: app/livechat/server/api/v1/visitor.ts [hunk 2]
```diff
@@ -37,13 +37,15 @@ API.v1.addRoute('livechat/visitor', {
 		}
 
 		guest.connectionData = normalizeHttpHeaderData(this.request.headers);
-		const visitorId = Livechat.registerGuest(guest);
+		const visitorId = Livechat.registerGuest(guest as any); // TODO: Rewrite Livechat to TS
```
Comment (KevLehman) [inline app/livechat/server/api/v1/visitor.ts:40]: Note: This TODO is Out of scope

File: client/providers/CallProvider/CallProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import { OutgoingByeRequest } from 'sip.js/lib/core';
+
+import { roomTypes } from '../../../app/utils/client';
+import { CallContext, CallContextValue } from '../../contexts/CallContext';
+import { useEndpoint } from '../../contexts/ServerContext';
+import { useUser } from '../../contexts/UserContext';
+import { isUseVoipClientResultError, isUseVoipClientResultLoading, useVoipClient } from './hooks/useVoipClient';
+
+export const CallProvider: FC = ({ children }) => {
+	// TODO: Test Settings and return false if its disabled (based on the settings)
```
Comment (KevLehman) [inline client/providers/CallProvider/CallProvider.tsx:13]: Is this part of MVP?

File: client/providers/CallProvider/CallProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	const result = useVoipClient();
+
+	const remoteAudioMediaRef = useRef<HTMLAudioElement>(null); // TODO: Create a dedicated file for the AUDIO and make the controls accessible
```
Comment (KevLehman) [inline client/providers/CallProvider/CallProvider.tsx:16]: Note: This TODO is Out of scope (Will be worked after MVP)

File: client/providers/CallProvider/hooks/useVoipClient.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+					}
+				})();
+			},
+			(error) => {
+				setResult({ error: error as Error });
+			},
+		);
+		return (): void => {
+			// client?.disconnect();
+			// TODO how to close the client? before creating a new one?
```
Comment (KevLehman) [inline client/providers/CallProvider/hooks/useVoipClient.ts:61]: is this part of MVP?

File: client/providers/OmnichannelProvider.tsx [hunk 6]
```diff
@@ -26,18 +28,20 @@ const OmnichannelProvider: FC = ({ children }) => {
...
 	const omnichannelPoolMaxIncoming = useSetting('Livechat_guest_pool_max_number_incoming_livechats_displayed') as number;
 
+	const loggerRef = useRef(new ClientLogger('OmnichannelProvider'));
 	const hasAccess = usePermission('view-l-room');
 	const canViewOmnichannelQueue = usePermission('view-livechat-queue');
-
 	const user = useUser() as IOmnichannelAgent;
 
 	const agentAvailable = user?.statusLivechat === 'available';
+	const voipCallAvailable = true; // TODO: use backend check;
```
Comment (KevLehman) [inline client/providers/OmnichannelProvider.tsx:37]: Is this part of MVP?

File: client/sidebar/sections/components/OmnichannelCallToggleReady.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+	const [registered, setRegistered] = useState(false);
+
+	const voipCallIcon = {
+		title: !registered ? t('Enable') : t('Disable'),
+		color: registered ? 'success' : undefined,
+		icon: registered ? 'phone' : 'phone-disabled',
+	} as const;
+	const voipClient = useCallClient();
+
+	// TODO: move registration flow to context provider
```
Comment (KevLehman) [inline client/sidebar/sections/components/OmnichannelCallToggleReady.tsx:20]: Note: This TODO is Out of scope (will be worked after MVP)

File: client/sidebar/sections/components/OmnichannelCallToggleReady.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	const handleVoipCallStatusChange = useMutableCallback((): void => {
+		// TODO: backend set voip call status
```
Comment (KevLehman) [inline client/sidebar/sections/components/OmnichannelCallToggleReady.tsx:22]: Is this part of MVP? (or even a valid comment 🙈 at this point?)

File: client/views/omnichannel/agents/AgentEdit.tsx [hunk 9]
```diff
@@ -1,41 +1,66 @@
...
 import { useMethod } from '../../../contexts/ServerContext';
+import { useSetting } from '../../../contexts/SettingsContext';
 import { useToastMessageDispatch } from '../../../contexts/ToastMessagesContext';
 import { useTranslation } from '../../../contexts/TranslationContext';
 import { useForm } from '../../../hooks/useForm';
 import UserInfo from '../../room/contextualBar/UserInfo';
 import { formsSubscription } from '../additionalForms';
 
-function AgentEdit({ data, userDepartments, availableDepartments, uid, reset, ...props }) {
+// TODO: TYPE:
```
Comment (KevLehman) [inline client/views/omnichannel/agents/AgentEdit.tsx:18]: 😬 this should be done before MVP if possible

File: client/views/room/Header/ToolBox/ToolBox.tsx [hunk 10]
```diff
@@ -1,6 +1,6 @@
 import { Menu, Option, MenuProps, Box } from '@rocket.chat/fuselage';
 import { useMutableCallback } from '@rocket.chat/fuselage-hooks';
-import React, { memo, ReactNode, useRef, ComponentProps, FC, ReactElement } from 'react';
+import React, { memo, ReactNode, useRef, ComponentProps, ReactElement } from 'react';
 // import tinykeys from 'tinykeys';
```
Comment (KevLehman) [inline client/views/room/Header/ToolBox/ToolBox.tsx:4]: Remove comment

File: client/sidebar/sections/components/OmnichannelCallToggleReady.tsx (hunk 8 above)
Comment (KevLehman) [reply client/sidebar/sections/components/OmnichannelCallToggleReady.tsx:22]: Not valid anymore

**Review by casalsgh — APPROVED** [review]
LGTM - All conflicts resolved; all PRs properly reviewed and approved.
